staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1193,14 +1207,17 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
 
 // Implements watch.Interface.
 func (c *cacheWatcher) Stop() {
-	c.forget()
+	c.forget(false)
 }
 
 // we rely on the fact that stopLocked is actually protected by Cacher.Lock()
 func (c *cacheWatcher) stopLocked() {
 	if !c.stopped {
 		c.stopped = true
-		close(c.done)
+		// stop without draining the input channel was requested.
```

Comment:
@p0lyn0mial - this is the only change comparing to your PR - I think it's fine:
- if we initially requested stopping without draining - then we're already done
- if we initially requested stopping with draining, then we will not close c.done channel
  but given that terminateWatchers will not hit this path anyway, it seems just fine to drop that (ignore the second stopLocked)

Reply:
I think we should bring back the previous version, a watcher can be terminated via the `Stop` method. I think the HTTP handlers do that.

Reply:
They do terminate (e.g. on timeout):
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L66

I thought it shouldn't be a problem because worst case we will loose a bit of work of processing something that noone will handle on the receiving end.
But there is one corner case if the c.result gets full and that gorotuine will be forever waiting on putting to the channel that is full. In such case, closing c.done channel fixes the problem.

I fixed that and extended the comment.

Reply:
okay, thx